harness/determined/_preemption.py

```diff
@@ -0,0 +1,142 @@
+import json
+import selectors
+import socket
+import threading
+from typing import Any, Optional
+
+from determined.common import api
+from determined.common.api import certs
+
+
+class _PreemptionWatcher(threading.Thread):
+    """
+    _PreemptionWatcher connects to the master and asynchronously waits for a preemption signal.
+
+    _PreemptionWatcher.should_preempt() is non-blocking (after the initial contact is made with the
+    master) and returns a bool indicating if a preemption signal has been received yet.
+
+    Example usage:
+
+    .. code:: python
+
+       with _PreemptionWatcher(master_url, trial_id, cert) as p:
+           print("started!")
+           for i in range(10):
+               if p.should_preempt():
+                   print("preempted!")
+                   break
+               print('no preemption yet, waiting...')
+               time.sleep(1)
+           else:
+               print('finished without preemption signal')
+    """
+
+    def __init__(self, master_url: str, trial_id: int, cert: certs.Cert) -> None:
+        self._master_url = master_url
+        self._trial_id = trial_id
+        self._cert = cert
+
+        # We will use a selector loop as a non-blocking way to stream through the requests library.
+        # We use a socketpair for a cross-platform way to wake up a selector loop mid-stream.
+        self._rsock = None  # type: Optional[socket.socket]
+        self._wsock = None  # type: Optional[socket.socket]
+
+        self._should_preempt = None  # type: Optional[bool]
+
+        self._cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        assert self._rsock is not None
+        assert self._wsock is not None
+
+        with api.get(
+            self._master_url,
+            f"/api/v1/trials/{self._trial_id}/signals/preemption",
+            stream=True,
+            cert=self._cert,
+        ) as r:
+            lines = r.iter_lines()
+
+            # The first response is always immediate, so we always block for it.  This is doubly
+            # important because we can only select on the HTTP connection socket _after_ we call
+            # next(lines) at least once.
+            for line in lines:
+                if not line:
+                    # Empty keepalive message.
+                    continue
+                break
+
+            j = json.loads(line)
```

Comment:
perhaps under some network failures, the json content may be "partial", and will fail `json.load`, so I'd handle it similarly to a network failure on the top `api.get`.